### PR TITLE
feat: use FOR loop instead of while on Iterators

### PR DESCRIPTION
**Pull request by msrivastav13**
### What issues does this PR fix or reference?

#<Insert GitHub Issue>

## The PR fulfills these requirements:

[ ] Tests for the proposed changes have been added/updated.
[ ] Code linting and formatting was performed.

### Functionality Before

The IterationRecipes were using while loop syntax for iteration. This is a syntax that is not so friendly compared to the FOR loop.

### Functionality After

The while loop syntax is not so elegant compared to regular FOR loops. Winter 24 added a nice quality of life improvement in Apex to be able to use the simplified and much familiar FOR loop instead of while syntax.

More details here (https://help.salesforce.com/s/articleView?id=release-notes.rn_apex_iterator_foreach.htm&release=246&type=5)

### CODE REVIEW

File: force-app/tests/Testing Recipes/PSGTestingRecipes.cls [hunk 1]
```diff
@@ -2,11 +2,11 @@
 private class PSGTestingRecipes {
     @isTest
     static void testRecalculatePSG() {
-        // get the PSG by name (may have been modified in deployment)
+        // get the PSG by name
         PermissionSetGroup psg = [
             SELECT Id, Status
             FROM PermissionSetGroup
-            WHERE DeveloperName = 'SalesPersona'
+            WHERE DeveloperName = 'Permission_Set_Group_for_testing'
```
Comment (msrivastav13) [inline force-app/tests/Testing Recipes/PSGTestingRecipes.cls:9]: @pozil I had to modify this code to fix a test failure.

Looks like winter 24 orgs removed a out-of-box permission set group `SalesPersona` this class was assuming. I created a new one so we do not depend on anything existing.

**Comment by msrivastav13** [comment]
@pozil Unsure why the build uses Summer 23 scratch orgs.

**Comment by pozil** [comment]
This is really weird, I don't understand why the prerelease branch isn't detected in this case whereas it's detected for your second PR. I'll run the prerelease PR manually but I'll keep an eye out for similar issues.

File: force-app/main/default/classes/Collection Recipes/IterationRecipes.cls [hunk 2]
```diff
@@ -14,11 +14,11 @@ public with sharing class IterationRecipes {
      **/
     public static Integer iterateOnAccountList(List<Account> accounts) {
         Integer employeeCount = 0;
-        Iterator<Account> it = accounts.iterator();
+        // Lists and Sets implement the Iterable interface in Apex
+        Iterable<Account> accountIterator = accounts;
```
Comment (docbill) [inline force-app/main/default/classes/Collection Recipes/IterationRecipes.cls:18]: There is a design pattern I like using that is better, if accounts can be null.

```
for(Iterator<Account> acctIterator = accounts?.iterator();
      true == accounts?.hasNext();
) {
     Account acct = acctIterator.next();
     ...
}
```

The advantage of this syntax is if instead you did something like:

```
if(accounts != null) {
  for(Account acct : accounts) {
      ...
  }
}
``` 

You have a significantly higher cognitive complexity score.

And if you do not do the null check at all, you are risking a run-time exception.

Comment (msrivastav13) [reply force-app/main/default/classes/Collection Recipes/IterationRecipes.cls:18]: @docbill Thank you for the review. Please feel free to submit a PR or raise an issue. Happy to take the contributions.

**Pull request merged**